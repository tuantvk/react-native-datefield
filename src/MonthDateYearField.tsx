import React, { useRef, useState } from 'react';
import { View, StyleSheet, Keyboard, TextInput } from 'react-native';
import {
  int,
  formatYearDigits,
  getOnlyNumber,
  getDateDefault,
  daysInMonth,
} from './utils';
import Input from './Input';
import type { DateFieldProps } from './types';

const defaultProps: DateFieldProps = {
  labelDate: 'Date',
  labelMonth: 'Month',
  labelYear: 'Year',
  editable: true,
};

const MonthDateYearField: React.FunctionComponent<DateFieldProps> = ({
  testID,
  containerStyle,
  styleInput,
  labelDate,
  labelMonth,
  labelYear,
  defaultValue,
  onSubmit,
  editable,
}) => {
  const refYear = useRef<TextInput>(null);
  const refDate = useRef<TextInput>(null);
  const [dateValue, setDateValue] = useState(getDateDefault(defaultValue));

  const onChangeDate = (value: string) => {
    let date = getOnlyNumber(int(value) > 31 ? '31' : value);
    setDateValue((prev) => ({ ...prev, date }));
    if (date.length === 2) {
      refYear.current?.focus();
    }
  };

  const onChangeMonth = (value: string) => {
    let month = getOnlyNumber(int(value) > 12 ? '12' : value);
    setDateValue((prev) => ({
      ...prev,
      month,
      date: daysInMonth(dateValue.date, month),
    }));
    if (month.length === 2) {
      refDate.current?.focus();
    }
  };

  const onChangeYear = (value: string) => {
    let current: number = new Date().getFullYear();
    let year = getOnlyNumber(int(value) > current ? current.toString() : value);
    setDateValue((prev) => ({ ...prev, year }));
    if (year.length === 4) {
      onSubmitEditing(year);
      Keyboard.dismiss();
    }
  };

  const onSubmitEditing = (year: string) => {
    onSubmit &&
      onSubmit(new Date(`${year}-${dateValue.month}-${dateValue.date}`));
  };

  const onBlur = () => {
    let current = { ...dateValue };
    if (current.date === '0') {
      current.date = '01';
    }
    if (current.month === '0') {
      current.month = '01';
    }
    if (daysInMonth(current.date, current.month) !== current.date) {
      current.date = daysInMonth(current.date, current.month);
    }
    if (current.year === '0') {
      current.year = `${formatYearDigits(new Date().getFullYear())}`;
    }
    if (current.year.length > 1 && current.year.length < 4) {
      current.year = `${formatYearDigits(int(current.year))}`;
    }
    setDateValue(current);
    onSubmitEditing(current.year);
  };

  return (
    <View {...{ testID }} style={[styles.container, containerStyle]}>
      <Input
        value={dateValue.month}
        placeholder={labelMonth}
        style={styleInput}
        onChangeText={onChangeMonth}
        onSubmitEditing={() => refDate.current?.focus()}
        {...{ editable, onBlur }}
      />
      <Input
        ref={refDate}
        value={dateValue.date}
        placeholder={labelDate}
        style={styleInput}
        onChangeText={onChangeDate}
        onSubmitEditing={() => refYear.current?.focus()}
        {...{ editable, onBlur }}
      />
      <Input
        ref={refYear}
        value={dateValue.year}
        maxLength={4}
        returnKeyType="done"
        placeholder={labelYear}
        style={styleInput}
        onChangeText={onChangeYear}
        onSubmitEditing={() => Keyboard.dismiss()}
        {...{ editable, onBlur }}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  input: {
    textAlign: 'center',
  },
});

MonthDateYearField.defaultProps = defaultProps;
export default MonthDateYearField;
