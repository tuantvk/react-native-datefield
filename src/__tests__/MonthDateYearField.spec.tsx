import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import MonthDateYearField from '../MonthDateYearField';

describe('MonthDateYearField component', () => {
  it('should match snapshot', () => {
    const component = shallow(<MonthDateYearField />);
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  describe('Props', () => {
    it('testID', () => {
      const component = shallow(<MonthDateYearField testID="date_input" />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('containerStyle', () => {
      const component = shallow(
        <MonthDateYearField containerStyle={{ marginVertical: 20 }} />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('styleInput', () => {
      const component = shallow(
        <MonthDateYearField styleInput={{ fontSize: 15 }} />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelDate', () => {
      const component = shallow(<MonthDateYearField labelDate="Input date" />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelMonth', () => {
      const component = shallow(
        <MonthDateYearField labelMonth="Input month" />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelYear', () => {
      const component = shallow(<MonthDateYearField labelYear="Input year" />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('defaultValue', () => {
      const component = shallow(
        <MonthDateYearField defaultValue={new Date('2021-03-25')} />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('onSubmit', () => {
      const component = shallow(<MonthDateYearField onSubmit={() => {}} />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('editable', () => {
      const component = shallow(<MonthDateYearField editable={false} />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('placeholderTextColor', () => {
      const component = shallow(
        <MonthDateYearField placeholderTextColor="#000000" />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
