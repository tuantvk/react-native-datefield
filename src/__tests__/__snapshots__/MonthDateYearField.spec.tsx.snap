// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthDateYearField component Props containerStyle 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      Object {
        "marginVertical": 20,
      },
    ]
  }
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
</View>
`;

exports[`MonthDateYearField component Props defaultValue 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    value="3"
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    value="25"
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value="2021"
  />
</View>
`;

exports[`MonthDateYearField component Props editable 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Memo(ForwardRef)
    editable={false}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    value=""
  />
  <Memo(ForwardRef)
    editable={false}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    value=""
  />
  <Memo(ForwardRef)
    editable={false}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
</View>
`;

exports[`MonthDateYearField component Props handleErrors 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
</View>
`;

exports[`MonthDateYearField component Props labelDate 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Input date"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
</View>
`;

exports[`MonthDateYearField component Props labelMonth 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Input month"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
</View>
`;

exports[`MonthDateYearField component Props labelYear 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Input year"
    returnKeyType="done"
    value=""
  />
</View>
`;

exports[`MonthDateYearField component Props maximumDate 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
</View>
`;

exports[`MonthDateYearField component Props minimumDate 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
</View>
`;

exports[`MonthDateYearField component Props onSubmit 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
</View>
`;

exports[`MonthDateYearField component Props placeholderTextColor 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    placeholderTextColor="#000000"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    placeholderTextColor="#000000"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    placeholderTextColor="#000000"
    returnKeyType="done"
    value=""
  />
</View>
`;

exports[`MonthDateYearField component Props styleInput 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    style={
      Object {
        "fontSize": 15,
      }
    }
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    style={
      Object {
        "fontSize": 15,
      }
    }
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    style={
      Object {
        "fontSize": 15,
      }
    }
    value=""
  />
</View>
`;

exports[`MonthDateYearField component Props testID 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
  testID="date_input"
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
</View>
`;

exports[`MonthDateYearField component should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    value=""
  />
  <Memo(ForwardRef)
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
</View>
`;
